File: rust/tests/listener_test.rs
use wgrok::create_listener;
use wgrok::get_logger;

#[test]
fn test_create_listener_factory_webex() {
    let logger = get_logger(false, "test");
    let result = create_listener("webex", "test_token_webex", logger);
    assert!(result.is_ok(), "Should create Webex listener");
}

#[test]
fn test_create_listener_factory_slack() {
    let logger = get_logger(false, "test");
    let result = create_listener("slack", "xapp-test-token", logger);
    assert!(result.is_ok(), "Should create Slack listener");
}

#[test]
fn test_create_listener_factory_discord() {
    let logger = get_logger(false, "test");
    let result = create_listener("discord", "test_token_discord", logger);
    assert!(result.is_ok(), "Should create Discord listener");
}

#[test]
fn test_create_listener_factory_irc() {
    let logger = get_logger(false, "test");
    let result = create_listener("irc", "nick@localhost:6667/channel", logger);
    assert!(result.is_ok(), "Should create IRC listener");
}

#[test]
fn test_create_listener_factory_invalid_platform() {
    let logger = get_logger(false, "test");
    let result = create_listener("invalid_platform", "token", logger);
    assert!(result.is_err(), "Should fail for invalid platform");
    match result {
        Err(e) => assert!(e.contains("Unsupported"), "Error should mention unsupported platform"),
        Ok(_) => panic!("Should have failed for invalid platform"),
    }
}

#[test]
fn test_incoming_message_structure() {
    use wgrok::IncomingMessage;
    use serde_json::json;

    let cards = vec![json!({"title": "Test Card"})];
    let msg = IncomingMessage {
        sender: "user@example.com".to_string(),
        text: "Hello, world!".to_string(),
        html: String::new(),
        msg_id: "msg_123".to_string(),
        platform: "webex".to_string(),
        cards,
    };

    assert_eq!(msg.sender, "user@example.com");
    assert_eq!(msg.text, "Hello, world!");
    assert_eq!(msg.msg_id, "msg_123");
    assert_eq!(msg.platform, "webex");
    assert_eq!(msg.cards.len(), 1);
}

#[test]
fn test_incoming_message_clone() {
    use wgrok::IncomingMessage;

    let msg1 = IncomingMessage {
        sender: "alice@example.com".to_string(),
        text: "Test message".to_string(),
        html: String::new(),
        msg_id: "id_1".to_string(),
        platform: "slack".to_string(),
        cards: vec![],
    };

    let msg2 = msg1.clone();
    assert_eq!(msg1.sender, msg2.sender);
    assert_eq!(msg1.platform, msg2.platform);
}

#[tokio::test]
async fn test_message_callback_mpsc() {
    use wgrok::IncomingMessage;
    use tokio::sync::mpsc;

    let (tx, mut rx) = mpsc::unbounded_channel();

    let msg = IncomingMessage {
        sender: "test@example.com".to_string(),
        text: "Test".to_string(),
        html: String::new(),
        msg_id: "123".to_string(),
        platform: "test".to_string(),
        cards: vec![],
    };

    tx.send(msg.clone()).expect("Should send message");

    let received = rx.recv().await;
    assert!(received.is_some(), "Should receive message");
    let received_msg = received.unwrap();
    assert_eq!(received_msg.sender, msg.sender);
    assert_eq!(received_msg.text, msg.text);
}

File: ts/tests/listener.test.ts
import {
  createListener,
  type IncomingMessage,
  WebexListener,
  SlackListener,
  DiscordListener,
  IrcListener,
} from '../src/listener';
import { noopLogger } from '../src/logging';

describe('listener factory', () => {
  it('creates WebexListener for webex platform', () => {
    const listener = createListener('webex', 'test-token', noopLogger);
    expect(listener).toBeInstanceOf(WebexListener);
  });

  it('creates SlackListener for slack platform', () => {
    const listener = createListener('slack', 'xapp-test-token', noopLogger);
    expect(listener).toBeInstanceOf(SlackListener);
  });

  it('creates DiscordListener for discord platform', () => {
    const listener = createListener('discord', 'test-token', noopLogger);
    expect(listener).toBeInstanceOf(DiscordListener);
  });

  it('creates IrcListener for irc platform', () => {
    const listener = createListener('irc', 'nick:pass@server:6697/#channel', noopLogger);
    expect(listener).toBeInstanceOf(IrcListener);
  });

  it('throws for unsupported platform', () => {
    expect(() => createListener('unknown', 'token', noopLogger)).toThrow('Unsupported platform: unknown');
  });
});

describe('IncomingMessage', () => {
  it('has required properties', () => {
    const msg: IncomingMessage = {
      sender: 'user@example.com',
      text: 'Hello world',
      html: '',
      roomId: '',
      roomType: '',
      msgId: 'msg-123',
      platform: 'webex',
      cards: [],
    };

    expect(msg.sender).toBe('user@example.com');
    expect(msg.text).toBe('Hello world');
    expect(msg.msgId).toBe('msg-123');
    expect(msg.platform).toBe('webex');
    expect(msg.cards).toEqual([]);
  });

  it('can hold card data', () => {
    const card = { type: 'AdaptiveCard', body: [] };
    const msg: IncomingMessage = {
      sender: 'user@example.com',
      text: 'Message with card',
      html: '',
      roomId: '',
      roomType: '',
      msgId: 'msg-456',
      platform: 'webex',
      cards: [card],
    };

    expect(msg.cards).toHaveLength(1);
    expect(msg.cards[0]).toBe(card);
  });
});
